### PR TITLE
fix(python): ensure `from_dicts` drops columns explicitly omitted from schema

### LINKED ISSUE #9570: Partial schema in from_dicts doesn't work with list schema
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

The docs say:

```
schema
...
The DataFrame schema may be declared in several ways:

As a dict of {name:type} pairs; if type is None, it will be auto-inferred.
As a list of column names; in this case types are automatically inferred.
...

If you want to drop some of the fields found in the input dictionaries, a _partial_ schema can be declared, in which case omitted fields will not be loaded. 
```

But a partial *list *schema does not drop the columns

### Reproducible example

```python
import polars as pl
pl.from_dicts([{"a": 1, "b": 2}], schema=["a"])
Out [39]:
shape: (1, 2)
┌─────┬─────┐
│ a   ┆ b   │
╞═════╪═════╡
│ 1   ┆ 2   │
└─────┴─────┘

pl.from_dicts([{"a": 1, "b": 2}], schema={"a": pl.Int8})
Out [40]:
shape: (1, 1)
┌─────┐
│ a   │
╞═════╡
│ 1   │
└─────┘
```

### Expected behavior

Both outputs should be identical

### Installed versions

<details>

```
--------Version info---------
Polars:      0.18.4
Index type:  UInt32
Platform:    Linux-6.2.6-76060206-generic-x86_64-with-glibc2.35
Python:      3.9.16 | packaged by conda-forge | (main, Feb  1 2023, 21:39:03) 
[GCC 11.3.0]

----Optional dependencies----
numpy:       1.24.2
pandas:      2.0.0
pyarrow:     11.0.0
connectorx:  0.3.1
deltalake:   <not installed>
fsspec:      2023.4.0
matplotlib:  3.7.1
xlsx2csv:    <not installed>
xlsxwriter:  <not installed>

```

</details>

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes #9570; ensures consistent behaviour between dict/list schema.

### CODE REVIEW

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
are you sure this would be correct for
```python
pl.from_dicts([{"a": 1, "b": 2}], schema=["d"])
```?

File: py-polars/polars/utils/_construction.py [hunk 1]
```diff
@@ -1036,8 +1036,13 @@ def _sequence_of_dict_to_pydf(
     )
     pydf = PyDataFrame.read_dicts(data, infer_schema_length, dicts_schema)
 
-    if column_names and set(column_names).intersection(pydf.columns()):
```
Comment (MarcoGorelli) [inline py-polars/polars/utils/_construction.py:1039]: for this section, would it work to just do
```python
    columns_equal = set(pydf.columns()) == set(column_names)
    if columns_equal and not schema_overrides:
        # skip unnecessary work
        pass
    elif column_names or schema_overrides:
        pydf = _post_apply_columns(
            pydf,
            columns=column_names,
            schema_overrides=schema_overrides,
        )
    return pydf
```
?

Comment (alexander-beedie) [reply py-polars/polars/utils/_construction.py:1039]: I like it :)

**Comment by alexander-beedie** [comment]
> are you sure this would be correct for

It is now, after your other suggestion ;)

**Comment by MarcoGorelli** [comment]
nice :) fancy including a test for `pl.from_dicts([{"a": 1, "b": 2}], schema=["d"])`, to check it raises and that this doesn't regress in the future?

**Comment by alexander-beedie** [comment]
> nice :) fancy including a test for `pl.from_dicts([{"a": 1, "b": 2}], schema=["d"])`, to check it raises and that this doesn't regress in the future?

Done; shame on me for not having already done so... ;)

**Review by MarcoGorelli — APPROVED** [review]
Nice one!

**Pull request merged**